From 85baf66840dbf728af732f1d346ffe63fbe3ab41 Mon Sep 17 00:00:00 2001
From: Jean-Yves Avenard <jyavenard@mythtv.org>
Date: Sun, 25 May 2014 16:36:05 +1000
Subject: [PATCH 2/8] Save volume only on TV Player exit

Original patch by angelaschmid <angela.schmid@wolke7.net>. Thank you

Fixes #12144
---
 mythtv/libs/libmythtv/audioplayer.cpp |    8 ++++++++
 mythtv/libs/libmythtv/audioplayer.h   |    1 +
 mythtv/libs/libmythtv/mythplayer.h    |    1 +
 mythtv/libs/libmythtv/tv_play.cpp     |    5 +++++
 4 files changed, 15 insertions(+)

diff --git a/mythtv/libs/libmythtv/audioplayer.cpp b/mythtv/libs/libmythtv/audioplayer.cpp
index 77dc7b8..ed40352 100644
--- a/mythtv/libs/libmythtv/audioplayer.cpp
+++ b/mythtv/libs/libmythtv/audioplayer.cpp
@@ -343,6 +343,14 @@ uint AudioPlayer::SetVolume(int newvolume)
     return GetVolume();
 }
 
+void AudioPlayer::SaveVolume(void)
+{
+    if (!m_audioOutput || m_no_audio_out)
+        return;
+    QMutexLocker lock(&m_lock);
+    m_audioOutput->SaveCurrentVolume();
+}
+
 int64_t AudioPlayer::GetAudioTime(void)
 {
     if (!m_audioOutput || m_no_audio_out)
diff --git a/mythtv/libs/libmythtv/audioplayer.h b/mythtv/libs/libmythtv/audioplayer.h
index eae034f..1efbf53 100644
--- a/mythtv/libs/libmythtv/audioplayer.h
+++ b/mythtv/libs/libmythtv/audioplayer.h
@@ -54,6 +54,7 @@ class MTV_PUBLIC AudioPlayer
     uint  GetVolume(void);
     uint  AdjustVolume(int change);
     uint  SetVolume(int newvolume);
+    void  SaveVolume(void);
     float GetStretchFactor(void) const { return m_stretchfactor; }
     void  SetStretchFactor(float factor);
     bool  IsUpmixing(void);
diff --git a/mythtv/libs/libmythtv/mythplayer.h b/mythtv/libs/libmythtv/mythplayer.h
index 541a0d5..70f3785 100644
--- a/mythtv/libs/libmythtv/mythplayer.h
+++ b/mythtv/libs/libmythtv/mythplayer.h
@@ -390,6 +390,7 @@ class MTV_PUBLIC MythPlayer
     // Audio Sets
     uint AdjustVolume(int change)           { return audio.AdjustVolume(change); }
     uint SetVolume(int newvolume)           { return audio.SetVolume(newvolume); }
+    void SaveVolume(void)                   { audio.SaveVolume(); }
     bool SetMuted(bool mute)                { return audio.SetMuted(mute);       }
     MuteState SetMuteState(MuteState state) { return audio.SetMuteState(state);  }
     MuteState IncrMuteState(void)           { return audio.IncrMuteState();      }
diff --git a/mythtv/libs/libmythtv/tv_play.cpp b/mythtv/libs/libmythtv/tv_play.cpp
index 0d4871e..e0726a5 100644
--- a/mythtv/libs/libmythtv/tv_play.cpp
+++ b/mythtv/libs/libmythtv/tv_play.cpp
@@ -3317,6 +3317,11 @@ void TV::PrepareToExitPlayer(PlayerContext *ctx, int line, BookmarkAction bookma
         }
         if (db_auto_set_watched)
             ctx->player->SetWatched();
+
+        if (ctx->player->GetAudio()->ControlsVolume())
+        {
+            ctx->player->SaveVolume();
+        }
     }
     ctx->UnlockDeletePlayer(__FILE__, line);
 }
-- 
1.7.10.2

